File: src/main/java/me/blake/pulsar/regions/CuboidUtil.java
package me.blake.pulsar.regions;

import org.bukkit.Location;

public class CuboidUtil {
    private final String worldName;
    private final int x1, y1, z1;
    private final int x2, y2, z2;

    public CuboidUtil(Location l1, Location l2) {
        if (!l1.getWorld().equals(l2.getWorld()))
            throw new IllegalArgumentException("Locations must be on the same world");
        this.worldName = l1.getWorld().getName();
        this.x1 = Math.min(l1.getBlockX(), l2.getBlockX());
        this.y1 = Math.min(l1.getBlockY(), l2.getBlockY());
        this.z1 = Math.min(l1.getBlockZ(), l2.getBlockZ());
        this.x2 = Math.max(l1.getBlockX(), l2.getBlockX());
        this.y2 = Math.max(l1.getBlockY(), l2.getBlockY());
        this.z2 = Math.max(l1.getBlockZ(), l2.getBlockZ());
    }

    private boolean contains(int x, int y, int z) {
        return x >= this.x1 && x <= this.x2 && y >= this.y1 && y <= this.y2 && z >= this.z1 && z <= this.z2;
    }

    public boolean contains(Location l) {
        if (!this.worldName.equals(l.getWorld().getName()))
            return false;
        return this.contains(l.getBlockX(), l.getBlockY(), l.getBlockZ());
    }

}

File: src/main/kotlin/me/blake/pulsar/region/CuboidUtil.kt
package me.blake.pulsar.region

import org.bukkit.Location

class CuboidUtil(
    l1: Location,
    l2: Location
) {
    private var worldName: String? = null
    private var x1 = 0; private var y1 = 0; private var z1 = 0
    private var x2 = 0; private var y2 = 0; private var z2 = 0

    init {
        require(l1.world == l2.world) { "Locations must be on the same world" }
        worldName = l1.world!!.name
        x1 = l1.blockX.coerceAtMost(l2.blockX)
        y1 = l1.blockY.coerceAtMost(l2.blockY)
        z1 = l1.blockZ.coerceAtMost(l2.blockZ)
        x2 = l1.blockX.coerceAtLeast(l2.blockX)
        y2 = l1.blockY.coerceAtLeast(l2.blockY)
        z2 = l1.blockZ.coerceAtLeast(l2.blockZ)
    }

    private fun contains(x: Int, y: Int, z: Int): Boolean {
        return x in x1..x2 && y >= y1 && y <= y2 && z >= z1 && z <= z2
    }

    operator fun contains(l: Location): Boolean {
        return if (worldName != l.world!!.name) false else this.contains(l.blockX, l.blockY, l.blockZ)
    }

}